### DISCUSSION TITLE
Do I need to return **all** refs from the setup store?

### DISCUSSION THREAD

**Discussion by ilyarefraction** [vuejs/pinia#2711] [Help and Questions] [answered]
As mentioned in the [docs](https://pinia.vuejs.org/core-concepts/#:~:text=Note%20that%20you%20must%20return%20all%20state%20properties%20in%20setup%20stores%20for%20Pinia%20to%20pick%20them%20up%20as%20state.%20In%20other%20words%2C%20you%20cannot%20have%20private%20state%20properties%20in%20stores.%20Not%20returning%20all%20state%20properties%20or%20making%20them%20readonly%20will%20break%20SSR%2C%20devtools%2C%20and%20other%20plugins.) we need to return all state properties, created with `ref()`, from a setup function.

It was not clear to me and my colleagues, whether this rules apply to _any_ `ref`'s created in the setup setup function, e.g. refs which are created with `storeToRefs`. 

<details>
<summary>Example</summary>

```js
const usePrivateUserRightsStore = defineStore(`private`, () => {
  const isAdmin = ref(false);

  return {
    isAdmin,
  };
});

export default defineStore(`public`, () => {
  const { isAdmin } = storeToRefs(usePrivateUserRightsStore()); // do I need to return `isAdmin` from this store?

  const checkIsAdmin = () => isAdmin.value;

  return {
    checkIsAdmin,
  };
});
```
</details>

For this discussion I have 3 question:
1. Do I need to return all actual `ref`'s from setup store?
2. Do you have examples/test when not returning state properties from the store break something?
3. Do you have explanation why this is important? As I can see from the source code this limitation most probably related to the [effectScrope](https://github.com/vuejs/pinia/blob/v2/packages/pinia/src/store.ts#L483) API but I didn't find any confirmation.

**Comment by posva** [comment, marked as answer]
1. Yes, all `ref` and `reactive` calls inside the store must be returned
2. This article should help: https://masteringpinia.com/blog/how-to-create-private-state-in-stores
3. The limitation does not come from effectScope. It's about pinia detecting state properties or not. The article explains the cases where this breaks


Regarding your example, any `storeToRefs()` implies an already existing reactive state, **so you don't return that**, it would duplicate the state in `pinia.state.value`

**Comment by Jiamingou** [comment]
But, from the document  "https://pinia.vuejs.org/cookbook/composing-stores.html#Nested-Stores" it demonstrated otherwise :

```js
import { defineStore } from 'pinia'
import { useUserStore } from './user'
import { apiPurchase } from './api'

export const useCartStore = defineStore('cart', () => {
  const user = useUserStore()
  const list = ref([])

  const summary = computed(() => {
    return `Hi ${user.name}, you have ${list.value.length} items in your cart. It costs ${price.value}.`
  })

  function purchase() {
    return apiPurchase(user.id, list.value)
  }

  return { summary, purchase }
})
````

the `list` state did not return.

**Reply by posva** [reply]
it's a typo, please submit a PR!

**Reply by Jiamingou** [reply]
done
